A cluster of changes to inference

ayush9pandey opened the pull request:
This is a PR to update inference module to fix various issues and add new features:

1. Fix #129 : The results of the inference run are written into `mcmc_results.csv` for the samples and `mcmc_results.txt` for the cost progress.
2. New feature as in #128 : Users can pass in model parameter arrays that condition the inference run when multiple trajectories are used.
3. Code cleanup: A lot of unused code was removed and other code cleanup in general.

Sorry for not managing this better and sending this all together. My branches got mixed at some point and it became hard to keep track. Also, please excuse the branch name.

ayush9pandey commented:
@WilliamIX Requesting to review. Also, since this is a bigger PR, it would be great if you could run this branch to make sure things are not breaking. 

I will also run code on my end to make sure things are fine.

WilliamIX commented on bioscrape/inference_setup.py:
```diff
@@ -168,35 +134,37 @@ def set_N_simulations(self, N_simulations: int):
+        A list of such dictionaries may be passed in 
+        corresponding to each measurement.
+        '''
+        self.parameter_conditions = parameter_conditions
```
Do we need any checks here to ensure that the parameter conditions are the correct size/shape? What about if the parameters being set are also parameters being inferred?

WilliamIX commented on bioscrape/inference_setup.py:
```diff
@@ -302,9 +270,26 @@ def extract_data(self, exp_data):
                 all_initial_conditions = self.initial_conditions
             self.initial_conditions = all_initial_conditions
+            # Create parameter conditions as required
+            if type(self.parameter_conditions) is dict:
```
Shouldn't these checks be in set_parameter_conditions?

ayush9pandey replied:
Since I am also checking the size of data list/dataframe passed to make sure the lengths of condition lists are correct, I am doing these checks later in `prepare_...` functions.

WilliamIX commented on bioscrape/inference_setup.py:
```diff
@@ -426,7 +425,8 @@ def run_emcee(self, **kwargs):
         skip_initial_state_check = kwargs.get('skip_initial_state_check', False)
         progress = kwargs.get('progess', True)
         threads = kwargs.get('threads', 1)
-        fname = kwargs.get('results_filename', 'mcmc_results.csv')
```
Can we leave this here for backwards compatibility, but maybe put a "Deprecation warning?"

ayush9pandey replied:
Done!

WilliamIX commented on bioscrape/inference_setup.py:
```diff
@@ -506,49 +512,33 @@ def plot_mcmc_results(self, sampler, plot_show = True, **kwargs):
-        if 'thin' in kwargs.keys():
-            thin = kwargs.get('thin')
-        else:
-            if convergence_check:
```
Have you removed all convergence checks? I personally found this feature annoying because it would just fail without returning anything so often a lot of computation would be wasted if convergence check were true.

ayush9pandey replied:
I didn't remove them but changed them quite a bit so that they are not hard-coded. Also, I set `convergence_check = False` to be the default since I also felt that it is annoying to lose the whole run just because emcee thinks it did not converge. User can now set `convergence_check = True` if they need to compute autocorrelation times in their MCMC runs.

WilliamIX reviewed (changes requested):
Left a couple of comments/questions in the review. I think also, considering the scope of all these changes, it would be very helpful to have some unit tests ensuring that the new features/functionality run (and to prevent me or someone else from breaking them later).

ayush9pandey commented:
Added unit tests for the new features. Also, confirming that I can run all inference examples with this branch without errors.

ayush9pandey commented:
Fixed all other comments and added unit tests. Thanks @WilliamIX  for your review. Please merge if everything looks ok now and create issues on things I should look at later to improve the inference module.

WilliamIX reviewed (commented):
Changes look great and tests are good. I tried downloading the branch and running everything and it worked on my computer as well.

One thing I noticed is that the file "birth_death_data_multiple_conditions.csv" gets rewritten every time tests are run - maybe we could add this to git ignore?
